shell/platform/android/io/flutter/view/AccessibilityViewEmbedder.java

```diff
@@ -0,0 +1,312 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.view;
+
+import android.graphics.Rect;
+import android.os.Build;
+import android.util.Log;
+import android.view.View;
+import android.view.accessibility.AccessibilityNodeInfo;
+import android.view.accessibility.AccessibilityNodeProvider;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Facilitates embedding of platform views in the accessibility tree generated by the accessibility bridge.
+ *
+ * Embedding is done by mirroring the accessibility tree of the platform view as a subtree of the flutter
+ * accessibility tree.
+ *
+ * This class relies on hidden system APIs to extract the accessibility information and does not work starting
+ * Android P; If the reflection accessors are not available we fail silently by embedding a null node, the app
+ * continues working but the accessibility information for the platform view will not be embedded.
+ *
+ * We use the term `flutterId` for virtual accessibility node IDs in the FlutterView tree, and the term `originId`
+ * for the virtual accessibility node IDs in the platform view's tree. Internally this class maintains a bidirectional
+ * mapping between `flutterId`s and the corresponding platform view and `originId`.
+ */
+class AccessibilityViewEmbedder {
+    private static final String TAG = "AccessibilityBridge";
+
+    private final ReflectionAccessors reflectionAccessors;
+
+    // The view to which the platform view is embedded, this is typically FlutterView.
+    private final View rootAccessibilityView;
+
+    // Maps a flutterId to the corresponding platform view and originId.
+    private final Map<Integer, ViewAndId> flutterIdToOrigin;
+
+    // Maps a platform view and originId to a corresponding flutterID.
+    private final Map<ViewAndId, Integer> originToFlutterId;
+
+    // Maps the flutterId of an accessibility node to the screen bounds of
+    // the root semantic node for the embedded view.
+    // This is used to translate the coordinates of the accessibility node subtree to the main display's coordinate
+    // system.
+    private final Map<Integer, Rect> flutterIdToDisplayBounds;
+
+    private int nextFlutterId;
+
+    AccessibilityViewEmbedder(View rootAccessibiiltyView, int firstVirtualNodeId) {
+        reflectionAccessors = new ReflectionAccessors();
+        flutterIdToOrigin = new HashMap<>();
+        this.rootAccessibilityView = rootAccessibiiltyView;
+        nextFlutterId = firstVirtualNodeId;
+        flutterIdToDisplayBounds = new HashMap<>();
+        originToFlutterId = new HashMap<>();
+    }
+
+    public AccessibilityNodeInfo getRootNode(View embeddedView, int flutterId, Rect displayBounds) {
```

Comment:
public api should have a doc comment

Reply:
Note that the class visibility is package private, I'm using "public" to make it clear which methods are used from within the package. Anyway I'll add some docs.

Reply:
package private, what a concept...